File: src/worker/game/rng.ts
// Deterministic random-generation engine (ported from Python `rng.py`).
//
// Per the GDD, every scenario, role and clue pull comes from a seedable PRNG —
// no AI / API token, ever. Seeding makes a whole game reproducible (and unit-
// testable), and lets a game be replayed from its short seed token.
//
// We use `mulberry32`: a tiny, fast, well-distributed 32-bit PRNG. Combined with
// a stable string->int hash (FNV-1a), the same seed token always yields the same
// stream regardless of host or run.

// Unambiguous alphabet: no 0/O, 1/I/L — easy to read/type as a room code.
const ROOM_CODE_ALPHABET = "ABCDEFGHJKMNPQRSTUVWXYZ23456789";
const SEED_TOKEN_ALPHABET = ROOM_CODE_ALPHABET;

/** A seedable random source for one game. */
export class GameRNG {
  readonly seed: string;
  private next: () => number; // returns a float in [0, 1)

  constructor(seed?: string) {
    this.seed = seed ?? makeSeedToken();
    this.next = mulberry32(fnv1a(this.seed));
  }

  /** Pick one item. Throws on an empty array (callers guard the banks). */
  choice<T>(items: readonly T[]): T {
    if (items.length === 0) throw new Error("cannot choose from an empty array");
    return items[Math.floor(this.next() * items.length)];
  }

  /** Pick `k` distinct items (partial Fisher–Yates over a copy). */
  sample<T>(items: readonly T[], k: number): T[] {
    if (k > items.length) throw new Error(`cannot sample ${k} from ${items.length}`);
    const pool = [...items];
    for (let i = 0; i < k; i++) {
      const j = i + Math.floor(this.next() * (pool.length - i));
      [pool[i], pool[j]] = [pool[j], pool[i]];
    }
    return pool.slice(0, k);
  }

  /** Return a new shuffled array; never mutates the input. */
  shuffled<T>(items: readonly T[]): T[] {
    const out = [...items];
    for (let i = out.length - 1; i > 0; i--) {
      const j = Math.floor(this.next() * (i + 1));
      [out[i], out[j]] = [out[j], out[i]];
    }
    return out;
  }

  /** Inclusive integer in [low, high]. */
  randint(low: number, high: number): number {
    return low + Math.floor(this.next() * (high - low + 1));
  }

  /**
   * The GDD's "intelligently loop through and choose the variation": a clue
   * scenario may map to many hand-written variations; this draws one. Named for
   * intent at call sites and so it can grow smarter later without touching them.
   */
  pickVariation<T>(options: readonly T[]): T {
    return this.choice(options);
  }

  /** A fresh 4-character room code from the unambiguous alphabet. */
  roomCode(): string {
    let code = "";
    for (let i = 0; i < 4; i++) code += ROOM_CODE_ALPHABET[Math.floor(this.next() * ROOM_CODE_ALPHABET.length)];
    return code;
  }
}

/**
 * Generate a fresh, human-readable seed token using the platform CSPRNG. The
 * token fully determines a `GameRNG` stream, so persisting it replays a game.
 */
export function makeSeedToken(length = 10): string {
  const bytes = new Uint8Array(length);
  crypto.getRandomValues(bytes);
  let out = "";
  for (const b of bytes) out += SEED_TOKEN_ALPHABET[b % SEED_TOKEN_ALPHABET.length];
  return out;
}

/** A stable 32-bit FNV-1a hash of `text` (deterministic across runs/hosts). */
function fnv1a(text: string): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < text.length; i++) {
    h ^= text.charCodeAt(i);
    // h *= 16777619, kept in 32-bit space.
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

/** mulberry32 PRNG: seed -> function returning floats in [0, 1). */
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

File: server/app/rng.py
"""Deterministic random-generation engine for Dead in the Water.

Per the Game Design Document (section 11), every scenario, role assignment and
clue pull is produced by a *Python random sequence generator* — no AI / API
token is ever used. This module is that generator.

Everything funnels through :class:`GameRNG`, a thin, seedable wrapper around the
standard-library :class:`random.Random`. Seeding it makes the whole game
reproducible, which is what lets the engine be unit-tested and lets a single
game be replayed from its seed token.

The "token" the design document refers to is a *seed token* — a short,
human-readable string that fully determines a game's randomness. It is produced
here algorithmically (``make_seed_token``); it is NOT an authentication token
and never leaves the server unless we explicitly want a game to be replayable.
"""

from __future__ import annotations

import random
import time
from typing import Iterable, Sequence, TypeVar

T = TypeVar("T")

# Room codes and seed tokens use an unambiguous alphabet: no 0/O, 1/I/L which
# are easy to misread on a phone when typing a 4-character room code.
ROOM_CODE_ALPHABET = "ABCDEFGHJKMNPQRSTUVWXYZ23456789"
SEED_TOKEN_ALPHABET = "ABCDEFGHJKMNPQRSTUVWXYZ23456789"


class GameRNG:
    """A seedable random source for one game.

    Construct with an explicit ``seed`` for reproducible games (tests, replays)
    or leave it ``None`` to draw a fresh seed token from the OS clock + entropy.
    The resolved seed is always available as :attr:`seed` so a game can be
    replayed later.
    """

    def __init__(self, seed: str | None = None):
        self.seed: str = seed or make_seed_token()
        # Hash the string seed to an int so equal tokens always yield the same
        # stream regardless of Python's per-process hash randomisation.
        self._random = random.Random(_stable_hash(self.seed))

    # -- primitives ---------------------------------------------------------
    def choice(self, items: Sequence[T]) -> T:
        """Pick one item. Raises on an empty sequence (callers guard banks)."""
        if not items:
            raise ValueError("cannot choose from an empty sequence")
        return self._random.choice(items)

    def sample(self, items: Sequence[T], k: int) -> list[T]:
        """Pick ``k`` distinct items, preserving no particular order."""
        if k > len(items):
            raise ValueError(f"cannot sample {k} items from {len(items)}")
        return self._random.sample(list(items), k)

    def shuffled(self, items: Iterable[T]) -> list[T]:
        """Return a new shuffled list; never mutates the input."""
        out = list(items)
        self._random.shuffle(out)
        return out

    def randint(self, low: int, high: int) -> int:
        """Inclusive integer in ``[low, high]``."""
        return self._random.randint(low, high)

    def weighted_choice(self, items: Sequence[T], weights: Sequence[float]) -> T:
        """Pick one item with the given relative weights."""
        if not items:
            raise ValueError("cannot choose from an empty sequence")
        return self._random.choices(items, weights=weights, k=1)[0]

    # -- domain helpers -----------------------------------------------------
    def pick_variation(self, options: Sequence[T]) -> T:
        """The GDD's "intelligently loop through and choose the variation".

        A clue scenario can map to many hand-written variations; this draws one
        of them. Kept as a named method so the intent reads clearly at call
        sites and so variation selection can grow smarter later (e.g. avoiding
        recently-used variations) without touching callers.
        """
        return self.choice(options)

    def room_code(self) -> str:
        """A fresh 4-character room code from the unambiguous alphabet."""
        return "".join(self._random.choice(ROOM_CODE_ALPHABET) for _ in range(4))


def make_seed_token(length: int = 10) -> str:
    """Generate a fresh, human-readable seed token.

    Mixes a high-resolution clock with the OS CSPRNG so two games created in the
    same millisecond still differ. The token fully determines a :class:`GameRNG`
    stream, so persisting it is enough to replay a game.
    """
    rng = random.Random(f"{time.time_ns()}-{random.SystemRandom().random()}")
    return "".join(rng.choice(SEED_TOKEN_ALPHABET) for _ in range(length))


def _stable_hash(text: str) -> int:
    """A process-stable 64-bit hash of ``text`` (FNV-1a).

    ``hash()`` is salted per-process, which would make a given seed token
    produce different games on every run. FNV-1a is tiny, dependency-free and
    deterministic, which is exactly what seeding needs.
    """
    h = 0xCBF29CE484222325
    for byte in text.encode("utf-8"):
        h ^= byte
        h = (h * 0x100000001B3) & 0xFFFFFFFFFFFFFFFF
    return h
